implementations/rust/ockam/ockam_command/src/error.rs

```diff
@@ -1,59 +1,158 @@
+use crate::version::Version;
+use crate::{exitcode, fmt_log, ExitCode};
 use colorful::Colorful;
-use std::fmt::{Debug, Display, Formatter};
 
-use crate::version::Version;
-use crate::{exitcode, fmt_err, ExitCode};
+use miette::miette;
+use miette::Diagnostic;
+use std::fmt::Debug;
 
 pub type Result<T> = std::result::Result<T, Error>;
 
-#[derive(Debug)]
-pub struct Error {
-    code: ExitCode,
-    description: String,
-    cause: Option<String>,
+#[derive(Debug, thiserror::Error, Diagnostic)]
+pub enum Error {
+    // ==== 4xx Errors =====
+    // Not Found
+    #[diagnostic(
+        code(OCK404),
+        help("Please check the spelling and try again"),
+        url("https://docs.ockam.io/errors/OCK404")
+    )]
+    #[error("Unable to find {resource} named {resource_name}")]
+    NotFound {
+        resource: String,
+        resource_name: String,
+    },
+
+    // Unauthorized
+    #[diagnostic(
+        code(OCK401),
+        help("Be sure you are enrolled to the project and have the correct permissions"),
+        url("https://docs.ockam.io/errors/OCK401")
+    )]
+    #[error("Unauthorized to operate on this project as {identity}")]
+    Unauthorized { identity: String },
+
+    // Conflict
+    #[diagnostic(
+        code(OCK409),
+        help("Be sure there are no other {resource}'s as {resource_name}"),
+        url("https://docs.ockam.io/errors/OCK409")
+    )]
+    #[error("Conflict with {resource} named {resource_name}")]
+    Conflict {
+        resource: String,
+        resource_name: String,
+    },
+    // ==== End 4xx Errors =====
+
+    // ==== 5xx Errors ====
+    // InternalError
+    #[diagnostic(
+        code(OCK500),
+        help("Please report this issue, with a copy of your logs, to https://github.com/build-trust/ockam/issues"),
+        url("https://docs.ockam.io/errors/OCK500")
+    )]
+    #[error("{error_message}")]
+    InteralError {
+        error_message: String,
+        exit_code: ExitCode,
+    },
+
+    // Unavailable
+    #[diagnostic(
+        code(OCK503),
+        help("Please wait a few minutes and try again or restart {resource:?} {resource_name:?}."),
+        url("https://docs.ockam.io/errors/OCK503")
+    )]
+    #[error("{resource} {resource_name} is unavailable")]
+    Unavailable {
+        resource: String,
+        resource_name: String,
+    },
+    // ==== End 5xx Errors ====
 }
 
 impl Error {
-    pub fn new(code: ExitCode, err: anyhow::Error) -> Self {
+    pub fn new(code: ExitCode, err: miette::ErrReport) -> Self {
         assert_ne!(code, 0, "Error's exit code can't be OK");
-        Self {
-            code,
-            description: err.to_string(),
-            cause: err.source().map(|s| s.to_string()),
+        Error::InteralError {
+            error_message: err.to_string(),
+            exit_code: code,
         }
     }
 
-    pub fn code(&self) -> ExitCode {
-        self.code
+    pub fn new_software_error(human_err: &str, inner_err_msg: &str) -> Self {
```

Comment:
I am not sure what `software` conveys here (I guess it relates to the exit code). If we eventually create an `InternalError` I would stick with `internal` in terms of vocabulary